### ISSUE TITLE
A2-Nets: Double Attention Networks

### ISSUE THREAD

**Issue by AtsukiOsanai** [AtsukiOsanai/cv_survey#10]
- [paper](https://arxiv.org/pdf/1810.11579.pdf)
- [PyTorch1](https://github.com/gjylt/DoubleAttentionNet), [PyTorch2](https://github.com/nguyenvo09/Double-Attention-Network)
- NeurlIPS2018

**Comment by AtsukiOsanai** [comment]
- long-range interactionを効率的に扱うattention modelの提案
- 通常，このようなlong-rangeの情報を取り込むためにはlayerをstackしてreceptive fieldを拡げる方法が取られるが，
この方法には以下のdrawbacksがある
  - 不要にモデルサイズを大きくし，計算コストの増大やオーバーフィットのリスクを増加させる
  - 離れた位置の情報が相互に取り込まれるには複数のforward, backward計算が必要となり，最適化の面で非効率的である
  - いくつかの層を経ることで実効的には遅れた情報となってしまう

**Comment by AtsukiOsanai** [comment]
アイデアのイメージ
<img width="582" alt="スクリーンショット 2021-01-05 22 16 26" src="https://user-images.githubusercontent.com/42481903/103650736-d9f66800-4fa3-11eb-8993-0b3a3cfeee81.png">

**Comment by AtsukiOsanai** [comment]
- Figure 1 (a)
  - Attention Mapsは画像中の特定の領域をハイライトするようなAttention
  - このAttention Mapを重み(softmaxをとる)に，特徴量をaggregateしたのがGlobal descriptors
    - サイズはinputのchannel-dim（m） x Attention Mapのdim（n）
    - これはnこのm次元ベクトルのbag (codebook)と捉えられる
  - Attention vectorはn個のbagから特徴を作るための重み付けを表す
- Figure 1 (b)
詳細はcodeを見た方が理解しやすい。いわゆるself-attention。
